### PR TITLE
UI Improvements for Locked Admin Questions

### PR DESCRIPTION
**Pull request by nospame**
## Product Description
Adds form tree indicators for question locked status, updates question property enabled state immediately based on locking/unlocking, and adds a "select all"-like lock and unlock for all questions contained in a group when it gets locked or unlocked. 

Not shown: removes "Edit Source XML" and "Edit Bulk Translations" options when a form contains any locked questions and the `edit_locked_questions` feature is false or absent.

https://github.com/user-attachments/assets/60d5d47c-4179-4f86-af6a-37133b84c172

## Technical Summary
https://dimagi.atlassian.net/browse/SAAS-19545
UI updates as noted in product description, as well as some small refactors for code readability and efficiency.

## Feature Flag
In HQ the `lock` plugin is enabled by LOCKED_ADMIN_QUESTIONS.

### Safety story
Tested functionality locally. The vast majority of changes here are limited to the lock plugin so any current issues would be constrained to only those with the feature flag enabled for testing purposes, with any remaining edge cases hopefully caught by QA.

### Automated test coverage
Automated testing is added or modified for the key functionality added in this PR.

### QA Plan
This feature will get QA as a whole before removing the feature flag and making it generally available.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -1981,6 +1966,8 @@ fn.setTreeValidationIcon = function (mug) {
     }
 };
 
+fn.setTreeExtraIcons = function (mug) {};
```
Comment (nospame) [inline src/core.js:1975]: I think I should either add a comment/docstring here or set up a basic data structure (since we expect to access `data.extraIcons`). Do reviewers have thoughts on one option or the other?

Comment (dannyroberts) [inline src/core.js:1975]: May be nice to add a comment here and above (where `obj.data.extraIcons` is accessed) to say this is a hook for a plugin to implement icons.

File: tests/formdesigner.lock.js [hunk 2]
```diff
@@ -130,18 +130,6 @@ describe("The Lock plugin", function() {
...
-            const groupWithNestedLock = getMug('/data/group_with_nested_lock');
-            assert(call('_hasLockedChildren', groupWithNestedLock),
-                "group with deeply nested locked question should have locked children");
-            const subgroup = getMug('/data/group_with_nested_lock/subgroup');
-            assert(call('_hasLockedChildren', subgroup),
-                "subgroup directly containing locked question should have locked children");
-            const groupNoLock = getMug('/data/group_no_lock');
-            assert.isFalse(call('_hasLockedChildren', groupNoLock),
-                "group without locked children should return false");
-        });
```
Comment (dannyroberts) [inline tests/formdesigner.lock.js:143]: Would be nice to get a clear explanation for why these tests are safe/necessary to delete.

File: src/lock.js [hunk 3]
```diff
@@ -46,6 +46,22 @@ $.vellum.plugin("lock", {}, {
...
+        const node = mug.ufid && this.jstree('get_node', mug.ufid);
+        if (!node) { return; }
+
+        let icon = null;
+        if (mug.p.locked) {
+            icon = treeLockIcon(mug);
+        }
+
+        if (node.data.extraIcons?.lock !== icon) {
+            node.data.extraIcons = node.data.extraIcons || {};
```
Comment (dannyroberts) [inline src/lock.js:83]: Not a strong preference, but consider initializing `node.data.extraIcons` at an earlier stage so that it is always present and set to `{}` or `{lock: null}` or whatever the proper empty value is.

File: src/lock.js [hunk 4]
```diff
@@ -46,6 +46,22 @@ $.vellum.plugin("lock", {}, {
         }
         this.__callOld();
     },
+    setTreeExtraIcons: function (mug) {
```
Comment (dannyroberts) [inline src/lock.js:72]: Stray thought, maybe `syncTreeExtraIcons` or something like that would be a more descriptive name. `set` to me sounds like it takes a value and sets something to that value.

File: src/lock.js [hunk 5]
```diff
@@ -138,8 +155,38 @@ function hasLockedChildren(mug) {
     );
 }
 
+function hasUnlockedChildren(mug) {
+    return mug.form.getChildren(mug).some(child =>
```
Comment (dannyroberts) [inline src/lock.js:203]: Probably not worth doing anything about this, but we had a quick discussion that if full tree search ever becomes a performance problem here, that maybe breadth-first instead of depth-first might be heuristically faster.

File: tests/formdesigner.lock.js [hunk 6]
```diff
@@ -214,6 +214,53 @@ describe("The Lock plugin", function() {
...
+        it("shows a lock icon on a locked group with no unlocked children", function () {
+            const icon = getLockIcon('/data/locked_group');
+            assert(icon, "expected icon on locked group");
+            assert.include(icon, 'fa-lock');
+            assert.notInclude(icon, 'fa-unlock');
+        });
+
+        it("shows an unlock icon on a locked group with unlocked children", function () {
+            const icon = getLockIcon('/data/locked_group_with_unlocked_children');
+            assert(icon, "expected lock icon on locked select");
```
Comment (dannyroberts) [inline tests/formdesigner.lock.js]: Evan pointed this out: should say "locked group"

File: src/lock.js [hunk 7]
```diff
@@ -92,6 +92,12 @@ $.vellum.plugin("lock", {}, {
                 mug.p.set(attr, value);
                 mug.form.getChildren(mug).forEach(child => setLockedFromParent(child));
                 _this.setTreeExtraIcons(mug);
+
+                if (!mug.form.isLoadingXForm) {
+                    if (mug.parentMug) {
+                        updateParentTreeIcons(mug.parentMug);
+                    }
```
Comment (dannyroberts) [inline src/lock.js]: Your choice, but I noticed you could replace these 3 lines with just

```js
updateParentTreeIcons(mug)
```

File: src/lock.js [hunk 8]
```diff
@@ -41,9 +41,10 @@ $.vellum.plugin("lock", {}, {
         setLockedFromParent(mug);
     },
     setTreeActions: function (mug) {
-        if (mug.p.locked) {
-            mug.options.canAddChoices = false;
+        if (!mug.options.hasOwnProperty('_originalCanAddChoices')) {
+            mug.options._originalCanAddChoices = mug.options.canAddChoices;
```
Comment (dannyroberts) [inline src/lock.js]: A comment might be helpful here. Basically `_originalCanAddChoices` is whether a question, forgetting whether it's locked or not, fundamentally has `.canAddChoices`, and we save it here while we kind of hijack `canAddChoices` to mean whether (due to a question being locked or otherwise) it should be displayed with the ability to add chocies.

File: src/lock.js [hunk 9]
```diff
@@ -94,6 +95,9 @@ $.vellum.plugin("lock", {}, {
                 _this.setTreeExtraIcons(mug);
 
                 if (!mug.form.isLoadingXForm) {
+                    if (_.contains(["Select", "MSelect"], mug.__className)) {
```
Comment (dannyroberts) [inline src/lock.js]: What's special about `["Select", "MSelect"]` is that these are questions that normally can take choices. Maybe make this a constant, and use in other places where you need to figure this out?

File: src/lock.js [hunk 10]
```diff
@@ -196,10 +198,12 @@ function hasUnlockedChildren(mug) {
     );
 }
 
-function setLockedFromParent(mug) {
-    if (mug.parentMug && mug.options.isControlOnly) {
-        mug.p.set('locked', mug.parentMug.p.locked);
-    }
+function setControlOnlyChildrenLocked(mug) {
```
Comment (dannyroberts) [inline src/lock.js]: Name is a bit confusing. Can you come up with a better one?

File: src/lock.js [hunk 11]
```diff
@@ -113,13 +113,16 @@ $.vellum.plugin("lock", {}, {
...
+                _this.setTreeExtraIcons(mug);
+
                 if (_.contains(["Select", "MSelect"], mug.__className)) {
                     setControlOnlyChildrenLocked(mug);
                 }
-                _this.setTreeExtraIcons(mug);
 
                 if (!mug.form.isLoadingXForm) {
-                    if (_.contains(["Select", "MSelect"], mug.__className)) {
+                    if (mug.__className === "Group") {
```
Comment (dannyroberts) [inline src/lock.js]: Does "Group" (here and below) capture every relevant container type? (Testing on the call, Evan said this might not cover repeat groups.)

Like my feedback for selects, probably worth making this list of container types a constant.

File: src/lock.js (hunk 10 above)
Comment (nospame) [reply src/lock.js]: I didn't manage to get farther than the very verbose `propagateLockedToControlOnlyChildren`, the reason being that this is useful to both select/multi-select questions with choices, and those that use lookup table references. That also came with a rename to clarify "children" vs "descendants" that I hope makes it somewhat clearer what's happening https://github.com/dimagi/Vellum/pull/1229/commits/22571501648c710a3f1ed77e4dbfea80248c54d2

File: src/lock.js (hunk 8 above)
Comment (nospame) [reply src/lock.js]: Updated to simply reference the class name to determine if we should touch `canAddChoices`. https://github.com/dimagi/Vellum/pull/1229/commits/70c6f937875d9c71cbc0b3b9f53e7ac07ac99c6a

File: src/lock.js (hunk 4 above)
Comment (nospame) [reply src/lock.js:72]: I agree in general, but `set` is also what we use for the preexisting `setTreeValidationIcon` (which would show an error icon in the same place) and I think existing naming is more consistent with that.

File: src/core.js (hunk 1 above)
Comment (nospame) [reply src/core.js:1975]: https://github.com/dimagi/Vellum/pull/1229/commits/f2560238284b0bfa0e612dff5f4229fbfc59a603

File: src/lock.js (hunk 11 above)
Comment (nospame) [reply src/lock.js]: Captured here https://github.com/dimagi/Vellum/pull/1229/commits/40f487d1158e561c1a608c6ecd3339c63fdec754

File: src/lock.js (hunk 5 above)
Comment (nospame) [reply src/lock.js:203]: Thank you for bringing this up. I don't think it's immediately clear that breadth-first would be faster here, but it does seem important to consider since forms can get pretty complex / nested and, like you said, if we start noticing any performance issues for heavy users of Locked Questions features.

File: tests/formdesigner.lock.js (hunk 2 above)
Comment (nospame) [reply tests/formdesigner.lock.js:143]: You raised a good point which is something I'll put into practice going forward, that the better way to do this would be to update the test to use the new functions by exposing them on the lock.js plugin to ensure it still passes -- _then_ remove it if necessary.

In this case the test was removed, yes because the functionality is no longer exposed by lock.js, but also because it's already implicitly tested through "disallows moving a group that contains a locked question", "disallows deleting a group that contains a locked question", and "disallows changing ID of a group that contains a locked question" - plus additional tests that were added in later commits for the `fa-lock` / `fa-unlock` icons.

File: src/lock.js (hunk 3 above)
Comment (nospame) [reply src/lock.js:83]: Since this is just the concern of `setTreeExtraIcons` I don't think I'd want to set it anywhere outside this function. One option would be to initialize it on the `core` stub method - that requires getting the node via jstree just to set an empty `{}` and not do anything with it, which doesn't feel super useful to me. The other option I see might be just moving this a few lines up out of the conditional. Do you think that would be better?

**Review by dannyroberts — APPROVED** [review]

**Review by dannyroberts — APPROVED** [review]

File: src/dataSourceWidgets.js [hunk 12]
```diff
@@ -206,7 +211,8 @@ function fixtureWidget(mug, options, labelText) {
                                 }
                             }
                         });
-                    }
+                    },
+                    !!widget.isDisabled(),
```
Comment (millerdev) [inline src/dataSourceWidgets.js:215]: nit: This argument is far away from the function name, and it takes a bit to parse that it's the third argument, not to mention what it does. What do you think of passing it in an `options` object?

`{isDisabled: !!widget.isDisabled()},`

which facilitates something more like keyword arguments. In case not obvious, it would change the signature of `getUIElementWithEditButton` to `($uiElem, editFn, options)`.
